Error using parametrize with default arguments

With pytest 3.4.0:
```python
import pytest

@pytest.mark.parametrize("something", [True, False])
def test_something(something=False):
  pass
```
You get the error:
```
../../base/Python.framework/Versions/2.7/lib/python2.7/site-packages/pluggy/__init__.py:617: in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
../../base/Python.framework/Versions/2.7/lib/python2.7/site-packages/pluggy/__init__.py:222: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
../../base/Python.framework/Versions/2.7/lib/python2.7/site-packages/pluggy/__init__.py:216: in <lambda>
    firstresult=hook.spec_opts.get('firstresult'),
../../base/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/python.py:200: in pytest_pycollect_makeitem
    res = list(collector._genfunctions(name, obj))
../../base/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/python.py:378: in _genfunctions
    self.ihook.pytest_generate_tests(metafunc=metafunc)
../../base/Python.framework/Versions/2.7/lib/python2.7/site-packages/pluggy/__init__.py:617: in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
../../base/Python.framework/Versions/2.7/lib/python2.7/site-packages/pluggy/__init__.py:222: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
../../base/Python.framework/Versions/2.7/lib/python2.7/site-packages/pluggy/__init__.py:216: in <lambda>
    firstresult=hook.spec_opts.get('firstresult'),
../../base/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/python.py:125: in pytest_generate_tests
    metafunc.parametrize(*marker.args, **marker.kwargs)
../../base/Python.framework/Versions/2.7/lib/python2.7/site-packages/_pytest/python.py:805: in parametrize
    self.function, name, arg))
E   ValueError: <function test_something at 0x10e4969b0> uses no argument 'something'
```
It seems pytest is separating the concept of normal python arguments and 'fixtures' here, only classifying only default-free arguments as potential fixtures - and then only considering 'fixtures' as valid targets for parametrisation - but the error message completely obfuscates this: by any regular python interpretation the function *definitely* has an argument 'something'. 

The documentation on this - https://docs.pytest.org/en/latest/parametrize.html doesn't show this as an example but also doesn't mention that it's a constraint; it's only through reading the pytest source that I could work out what the problem was.

Related: The wording of this was changed to 'argument' from 'fixture' in https://github.com/pytest-dev/pytest/issues/1539 which makes sense in the context of that error, but only confuses this one.

--- a/_pytest/compat.py
+++ b/_pytest/compat.py
@@ -135,6 +135,14 @@ def getfuncargnames(function, is_method=False, cls=None):
     return arg_names
 
 
+def get_default_arg_names(function):
+    # Note: this code intentionally mirrors the code at the beginning of getfuncargnames,
+    # to get the arguments which were excluded from its result because they had default values
+    return tuple(p.name for p in signature(function).parameters.values()
+                 if p.kind in (Parameter.POSITIONAL_OR_KEYWORD, Parameter.KEYWORD_ONLY) and
+                 p.default is not Parameter.empty)
+
+
 if _PY3:
     STRING_TYPES = bytes, str
     UNICODE_TYPES = str,

--- a/_pytest/python.py
+++ b/_pytest/python.py
@@ -25,7 +25,7 @@ from _pytest.compat import (
     isclass, isfunction, is_generator, ascii_escaped,
     REGEX_TYPE, STRING_TYPES, NoneType, NOTSET,
     get_real_func, getfslineno, safe_getattr,
-    safe_str, getlocation, enum,
+    safe_str, getlocation, enum, get_default_arg_names
 )
 from _pytest.outcomes import fail
 from _pytest.mark.structures import transfer_markers
@@ -789,6 +789,7 @@ class Metafunc(fixtures.FuncargnamesCompatAttr):
         argnames, parameters = ParameterSet._for_parametrize(
             argnames, argvalues, self.function, self.config)
         del argvalues
+        default_arg_names = set(get_default_arg_names(self.function))
 
         if scope is None:
             scope = _find_parametrized_scope(argnames, self._arg2fixturedefs, indirect)
@@ -797,13 +798,16 @@ class Metafunc(fixtures.FuncargnamesCompatAttr):
         valtypes = {}
         for arg in argnames:
             if arg not in self.fixturenames:
-                if isinstance(indirect, (tuple, list)):
-                    name = 'fixture' if arg in indirect else 'argument'
+                if arg in default_arg_names:
+                    raise ValueError("%r already takes an argument %r with a default value" % (self.function, arg))
                 else:
-                    name = 'fixture' if indirect else 'argument'
-                raise ValueError(
-                    "%r uses no %s %r" % (
-                        self.function, name, arg))
+                    if isinstance(indirect, (tuple, list)):
+                        name = 'fixture' if arg in indirect else 'argument'
+                    else:
+                        name = 'fixture' if indirect else 'argument'
+                    raise ValueError(
+                        "%r uses no %s %r" % (
+                            self.function, name, arg))
 
         if indirect is True:
             valtypes = dict.fromkeys(argnames, "params")

--- a/changelog/3221.feature
+++ b/changelog/3221.feature
@@ -1,0 +1,1 @@
+Added a more indicative error message when parametrizing a function whose argument takes a default value.

--- a/testing/python/metafunc.py
+++ b/testing/python/metafunc.py
@@ -616,6 +616,19 @@ class TestMetafunc(object):
             "*uses no argument 'y'*",
         ])
 
+    def test_parametrize_gives_indicative_error_on_function_with_default_argument(self, testdir):
+        testdir.makepyfile("""
+            import pytest
+
+            @pytest.mark.parametrize('x, y', [('a', 'b')])
+            def test_simple(x, y=1):
+                assert len(x) == 1
+        """)
+        result = testdir.runpytest("--collect-only")
+        result.stdout.fnmatch_lines([
+            "*already takes an argument 'y' with a default value",
+        ])
+
     def test_addcalls_and_parametrize_indirect(self):
         def func(x, y):
             pass

